src/crc.rs

```diff
@@ -0,0 +1,356 @@
+            reg: crc,
+            output_xor: 0,
+        }
+    }
+
+    /// Set the unit's configuration, discarding previous state.
+    pub fn set_config(&mut self, config: &Config) {
+        self.output_xor = config.output_xor & config.poly.xor_mask();
+
+        // manual says unit must be reset (or DR read) before change of polynomial
+        // (technically only in case of ongoing calculation, but DR is buffered)
+        self.reg.cr.modify(|_, w| {
+            w.polysize()
+                .bits(config.poly.polysize())
+                .rev_in()
+                .bits(config.get_reverse_input())
+                .rev_out()
+                .bit(config.reverse_output)
+                .reset()
+                .set_bit()
+        });
+        self.reg.pol.write(|w| w.pol().bits(config.poly.pol()));
+        self.reg.init.write(|w| w.init().bits(config.initial));
+        // writing to INIT sets DR to its value
+    }
+
+    /// Write data to the CRC unit. Note that CRC calculation works
+    /// faster if more data is given at once.
+    pub fn update(&mut self, data: &[u8]) {
+        // write 4 bytes at once, then 2, then 1, as appropriate
+        // in the case of a single large slice this improves speed by >3x
+        let mut words = data.chunks_exact(4);
+        for word in words.by_ref() {
+            let word = u32::from_be_bytes(word.try_into().unwrap());
+            self.reg.dr_mut().write(|w| w.dr().bits(word));
+        }
+
+        // there will be at most 3 bytes remaining, so 1 half-word and 1 byte
+        let mut half_word = words.remainder().chunks_exact(2);
+        if let Some(half_word) = half_word.next() {
+            let half_word = u16::from_be_bytes(half_word.try_into().unwrap());
+            self.reg.dr16_mut().write(|w| w.dr16().bits(half_word));
+        }
+
+        if let Some(byte) = half_word.remainder().first() {
+            self.reg.dr8_mut().write(|w| w.dr8().bits(*byte));
+        }
+    }
+
+    /// Write data to the CRC unit, return CRC so far. This function should
+    /// only be used if you need its result, as retrieving the CRC takes time.
+    #[must_use = "retrieving the CRC takes time, use update() if not needed"]
+    pub fn update_and_read(&mut self, data: &[u8]) -> u32 {
+        self.update(data);
+        self.read_crc()
+    }
+
+    /// Read the CRC and reset DR to initial value in preparation for a new CRC.
+    /// This does not reset the configuration options.
+    pub fn finish(&mut self) -> u32 {
+        let result = self.read_crc();
+        self.reg.cr.modify(|_, w| w.reset().set_bit());
+        result
+    }
+
+    /// Read the CRC without resetting the unit.
+    #[inline]
+    pub fn read_crc(&self) -> u32 {
+        self.reg.dr().read().dr().bits() ^ self.output_xor
+    }
+
+    /// Write the independent data register. The IDR can be used as
+    /// temporary storage. It is not cleared on CRC hash reset.
+    pub fn set_idr(&mut self, value: u32) {
```

Comment:
Having an example of usage for this would be nice if you can come up with some.

Reply:
I don't know any specific use case beyond storing a u32. In many other chips it has 24 reserved bits, so it's possible it was used internally by the unit and just exposed because earlier chips had much less RAM.

It's possible to use it to store an ongoing CRC calculation* but it's not specific to the IDR (anywhere works) and mildly complicated to the point where I'd probably want to add assistance to the CRC impl itself, rather than have users do most of it. Some kind of `read_crc_state` method, eg:

```rust
pub fn read_crc_state(&self) -> u32 {
    let state = self.read_crc_no_xor();
    if self.reg.cr.read().rev_out().is_reversed() {
        state.reverse_bits()
    } else {
        state
    }
}
```

\* For example an incomplete CRC calculation could have its DR read and stored, then unit could be reset and used to calculate a different CRC for presumably less data, then config restored with old DR as INIT. I think the CRC unit supports DMA in some way, where this might be more useful. I've never used DMA on any MCU so I wouldn't know for sure.

Rough code example:

```rust
fn idr_example(crc: &mut Crc) {
    let config = Config::new().reflect(true).output_xor(0xFFFF_FFFF);
    crc.set_config(&config);
    crc.update(b"12345");
    crc.set_idr(crc.read_crc_state()); // store ongoing CRC

    crc.set_config(&Config::new()); // perform unrelated CRC
    assert_eq!(crc.update_and_read(CHECK_DATA), 0x0376_E6E7);

    // retrieve ongoing CRC from IDR, set it as the initial value
    crc.set_config(&config.initial_value(crc.get_idr()));
    assert_eq!(crc.update_and_read(b"6789"), 0xCBF4_3926); // finish CRC, check result
}
```

If this seems potentially useful I can add it. The `read_crc_state` method may be useful enough on its own without involving the IDR.

Reply:
Hmmm the reference manual doesn't tell us more than
```
The CRC_IDR register can be used to hold a temporary value related to CRC calculation.
```
either. And not even the CRC application note (AN4187), which does go into a bit of detail regarding DMA, mentions it. So its probably only some zombie register that nobody really needs I guess.

In this case, could you add a comment that explains to the user that this register doesn't play any role in the actual calculation of the CRC so people who accidentally stumble over it don't happen to waste their time on trying to figure out what it's good for?